command/agent/consul/check_watcher.go

```diff
@@ -0,0 +1,278 @@
+package consul
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/consul/api"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+const (
+	// defaultPollFreq is the default rate to poll the Consul Checks API
+	defaultPollFreq = 900 * time.Millisecond
+)
+
+// ChecksAPI is the part of the Consul API the checkWatcher requires.
+type ChecksAPI interface {
+	// Checks returns a list of all checks.
+	Checks() (map[string]*api.AgentCheck, error)
+}
+
+// TaskRestarter allows the checkWatcher to restart tasks and how long the
+// grace period should be afterward.
+type TaskRestarter interface {
+	RestartDelay() time.Duration
+	Restart(source, reason string, failure bool)
+}
+
+// checkRestart handles restarting a task if a check is unhealthy.
+type checkRestart struct {
+	allocID   string
+	taskName  string
+	checkID   string
+	checkName string
+
+	// remove this checkID (if true only checkID will be set)
+	remove bool
+
+	task           TaskRestarter
+	restartDelay   time.Duration
+	grace          time.Duration
+	interval       time.Duration
+	timeLimit      time.Duration
+	ignoreWarnings bool
+
+	// Mutable fields
+
+	// unhealthyStart is the time a check first went unhealthy. Set to the
+	// zero value if the check passes before timeLimit.
+	unhealthyStart time.Time
+
+	// graceUntil is when the check's grace period expires and unhealthy
+	// checks should be counted.
+	graceUntil time.Time
+
+	logger *log.Logger
+}
+
+// update restart state for check and restart task if necessary. Currrent
+// timestamp is passed in so all check updates have the same view of time (and
+// to ease testing).
+func (c *checkRestart) update(now time.Time, status string) {
+	healthy := func() {
+		if !c.unhealthyStart.IsZero() {
+			c.logger.Printf("[DEBUG] consul.health: alloc %q task %q check %q became healthy; canceling restart",
+				c.allocID, c.taskName, c.checkName)
+			c.unhealthyStart = time.Time{}
+		}
+		return
+	}
+	switch status {
+	case api.HealthCritical:
+	case api.HealthWarning:
+		if c.ignoreWarnings {
+			// Warnings are ignored, reset state and exit
+			healthy()
+			return
+		}
+	default:
+		// All other statuses are ok, reset state and exit
+		healthy()
+		return
+	}
+
+	if now.Before(c.graceUntil) {
+		// In grace period exit
+		return
+	}
+
+	if c.unhealthyStart.IsZero() {
+		// First failure, set restart deadline
+		if c.timeLimit != 0 {
+			c.logger.Printf("[DEBUG] consul.health: alloc %q task %q check %q became unhealthy. Restarting in %s if not healthy",
+				c.allocID, c.taskName, c.checkName, c.timeLimit)
+		}
+		c.unhealthyStart = now
```

Comment:
Can you rename unhealthyStart to unhealthyCheck?

Reply:
At no other point do I refer to a single run of a check as a check in this file, so I'm not sure I understand the renaming. We don't even have visibility into individual runs of a check since we poll statuses at a fixed interval.